### PR TITLE
Combines internal and external bleeding modifiers

### PR DESCRIPTION
**Pull request by jwhitak**
This PR combines the internal and external bleeding modifiers into a single helper function in order to simply the effects of blood loss for the medbay crew. It's difficult to remember what modifier affects what type of bleeding, so this simplifies all current effects into one place. This does *not* prevent specific modifiers from being added or changed later (ex, newChem1 that stops only external bleeding can be added safely).

This also affects some alternative effects that cause blood loss (anything that calls drip()) such as ruptured lungs.

Before PR:

> **EXTERNAL** (blood drops on ground, can fix with bandaids)
> lying down slows bleeding 30%
> hyperzine increases bleeding 30%
> inaprovaline slows bleeding 30%
> bicaridine does not affect external bleeding
> clotting agent/biofoam do not affect external bleeding
> cryo does not affect external bleeding

> **INTERNAL** (no visible blood drops, INTENRAL BLEEDING DETECTED, can fix with surgery)
> lying down does not affect internal bleeding
> hyperzine increases bleeding 30%
> inaprovaline and bicaridine slow bleeding 30% but do not stack with each other
> clotting agent/biofoam stop internal bleeding
> cryo stops internal bleeding

After PR:
> **BOTH EXTERNAL AND INTERNAL** (external still cured only by bandaids, internal still only by surgery)
> lying down slows bleeding 30%
> hyperzine increases bleeding 30%
> inaprov/bicard slow bleeding 30% and do not stack
> clotting agent/biofoam stop bleeding
> cryo slops bleeding

Basic testing is finished on monkeymen on a local server, nothing seems to be broken (famous last words).

[balance]

:cl:
 * rscadd: Combines internal and external bleeding modifiers.

### CODE REVIEW

**Comment by Exxion** [comment]
I definitely agree with most of this, but I'm not sure about biofoam stopping external bleeding. But that's kind of a nitpick.
Also internal bleeding isn't treatable *only* with surgery, but it's the fastest method

**Comment by jwhitak** [comment]
Bicaridine overdoses are forbidden knowledge. Did you know they also heal burns?

Biofoam/clotting agent/both can be moved to internal only if needed, depending on comments.
Edit: I do like keeping the stuff the same for both. It helps the non medbrains out in an emergency. Imagine, seeing your bro [BLEEDING] but you don't know what to do. But you heard biofoam stops bleeding... That said, this is still a medical buff and needs discussion.

**Comment by Exxion** [comment]
I can't remember if punching yourself to heal larger wounds was fixed or not, but that one was/is pretty good forbidden knowledge.

File: code/modules/organs/organ_external.dm [hunk 1]
```diff
@@ -610,18 +610,9 @@ Note that amputating the affected organ does in fact remove the infection from t
...
-			if(owner.reagents.has_reagent(HYPERZINE)) //On the other hand, hyperzine speeds up bleeding by 30%
-				blood_factor += 0.3
-
-			if(owner.reagents.has_reagent(CLOTTING_AGENT) || owner.reagents.has_reagent(BIOFOAM)) //Clotting agent and biofoam stop bleeding entirely.
-				blood_factor = 0
+		//Internal wounds get worse over time, and you lose blood
+		if(W.internal && !W.is_treated() && !(species && species.anatomy_flags & NO_BLOOD))
+			var/blood_factor = owner.calcbloodloss() //Calls helper function to determine amount of blood loss.
 
 			owner.vessel.remove_reagent(BLOOD, W.damage * wound_update_accuracy * 0.07 * max(0, blood_factor))
```
Comment (Kammerjunk) [inline code/modules/organs/organ_external.dm:617]: I was gonna suggest that this be done in an unrelated PR, but since these reagents checks are getting removed, it seems appropriate for this scope to re-implement wounds slowly opening more and more.
see [these three removed lines](https://github.com/vgstation-coders/vgstation13/pull/28915/files#diff-a1ca7744b58972eda50448593079d96db87072abec59d0d5780bcc99219d9c35L617) for how it used to work before hyperzine-bleeding was added.

Comment (Kammerjunk) [reply code/modules/organs/organ_external.dm:617]: and idk how necessary or in-scope this is, but it might also be nice to just have blood_factor be a variable on wounds, and then have these three lines instead read `W.blood_factor`
```
for(var/datum/wound/W in temp.wounds)
	if(W.bleeding())
		blood_max += W.damage / 4
```
in however a manner you choose, I suppose - whether they add to `blood_factor` or `blood_max` in handle_blood()

---
or in other words, something like this:

`code/modules/organs/wound.dm`
```
/datum/wound
  var/blood_factor = 1
```
`code/modules/organs/organ_external.dm`
```
/datum/organ/external/proc/update_wounds()
	...
	for(var/datum/wound/W in wounds)
		if(W.internal && !W.is_treated() && !(species && species.anatomy_flags & NO_BLOOD))
			blood_factor = owner.calcbloodloss() //Calls helper function to determine amount of blood loss.
		...
		// Wound-opening thing from comment above.
		if(!owner.reagents.has_any_reagents(list(BICARIDINE,INAPROVALINE,CLOTTING_AGENT,BIOFOAM)))	//Bicard, inaprovaline, clotting agent, and biofoam stop internal wounds from growing bigger with time, and also slow bleeding
			W.open_wound(0.1 * wound_update_accuracy)
			owner.vessel.remove_reagent(BLOOD, 0.05 * W.damage * wound_update_accuracy)
```
`code/modules/organs/blood.dm`
```
/mob/living/carbon/human/proc/handle_blood()
	...
	for(var/datum/wound/W in temp.wounds)
		if(W.bleeding())
			blood_max += W.blood_factor
```
or whatever, idk. would just be nice if there weren't 2 sources of on-Life bloodloss, or if the other one was only for wounds opening.

Comment (jwhitak) [reply code/modules/organs/organ_external.dm:617]: Response to your first comment is that it would be a different PR, it's a bit unatomic to unnerf something like the wounds-not-actually-expanding here.

Second comment's response is that wound code is really jank code. I don't want to go in depth here but:

> **blood.dm**
> handles all externally bleeding wounds
> bleeding() returns 0 if the wound is internal, so they aren't handled here at all

> **organ_external.dm**
> handles all "internal wounds" (wounds inside of external organs vs outside of external organs)
> internal wounds are all called "internal bleeding" but do not bleed normally as above
> internal wounds bleed as they expand
> due to an older PR, wounds no longer actually expand, but the blood loss is still there

>would just be nice if there weren't 2 sources of on-Life bloodloss, or if the other one was only for wounds opening.

Wounds opening are the only way internal wounds bleed, and there's only one blood.dm proc that causes bleeding (for external wounds). So, basically, this is already done.

Also, blood_max is maximum amount of blood loss (in units) added from ALL external bleeding wounds, blood_factor is the scaling of this blood loss based on the factors listed in the PR (so your blood loss formula would be something like bleed(blood_max * blood_factor). This PR basically organizes the blood_factor calculations into a single function that's only called when bleeding occurs. Saving the result in the wound itself is a waste of memory since the result can change every life() tick but only needs to be calculated when bleeding is happening. Same with blood_max, which is technically just a derivative calculation based on w.damage across all currently existing wounds on a mob's external organ in the first place.

I think we're good here, minus a refactoring of wound code, and the above re-adding of the expanding wounds code.

Comment (Kammerjunk) [reply code/modules/organs/organ_external.dm:617]: yeah but it _isn't_ in a single function. bloodloss happens in `handle_blood()` due to the eventual drip() call, and it also currently happens in update_wounds() due to the remove_reagent() call. they're distinct and separate for no good reason I can see, and the latter straight up just removes a volume of `BLOOD` from the vessel.
the reason the latter existed in the first place _was_ because of open_wounds(), but open_wounds() was removed and remove_reagent() is still there.

Comment (jwhitak) [reply code/modules/organs/organ_external.dm:617]: Handled on Discord(tm)

Tl;dr update_wounds() ONLY handles internal wounds.

**Comment by adacovsk** [comment]
This PR makes sense and is good.

However, biofoam is so rarely used, I've only ever used it as an antagonist, as people usually die before treatment is utilized, unless it's a lung rupture. Besides that, you'll probably be in medbay anyway so your bleeding will be fixed and biofoam won't be used.

IMO there should probably be a cap on bloodloss if there isn't, invap or bicard don't do anything at higher blood loss from my experience

**Comment by jwhitak** [comment]
30% reduction doesn't mean much when you are going to die in 3 seconds. Wow, 1 whole extra second of living! It probably took longer to inject the bicard into them. But this PR's goal isn't to straight up buff blood loss prevention, it's to simplify all sources of it (which is a buff by itself). Cryo should now properly stop the bleeding regardless of type, biofoam isn't needed so much in medbay because of this so parameds are more likely to use it in the field, etc.

As an aside, this PR affects all sources of drip(), including lung ruptures. While they would still have difficulties breathing, the blood spewing out of their lungs can be slowed/stopped with the standard chemicals/resting/cryo now.

**Comment by adacovsk** [comment]
> 30% reduction doesn't mean much when you are going to die in 3 seconds

My point is that even swallowing like 100u of bicard rarely ever stops IB longterm and you end up dying anyway even if you take antitox with it

**Comment by Eneocho** [comment]
> > 30% reduction doesn't mean much when you are going to die in 3 seconds
> 
> My point is that even swallowing like 100u of bicard rarely ever stops IB longterm and you end up dying anyway even if you take antitox with it

I've seen people treat IB by taking a fuckhueg dose of bicard and antitox, then a decent amount of iron, nutriment and dexalinplus. It eventually closes, but it takes like 10m.

**Comment by pedr69** [comment]
i'd make inaprov slow bleeding even more; a syringe of it is inside every medkit and always gets ignored for the DD injector
imo injecting it into a horizontal beaten up spaceman should make him take a fair bit longer to die
but that's probably something for another PR

**Comment by Eneocho** [comment]
I'd make inaprov and bicaridine stack, so the turbo prepared paramedic can get the space ninja victim to safety without em bleeding out impossibly fast.

**Comment by jwhitak** [comment]
Yeah, changing the actual balance of the modifiers can come later. Also, inaprov is not only slightly useless, but it's actual poison to vox, so most docs tend to ignore it. It could possibly use a buff, and bleeding might be a good target. This is also a pretty open area for newer, fun chemicals to enter into the game. For example, more things that could increase bleeding.

**Comment by Eneocho** [comment]
I was thinking of adding a chem to garlic to increase bleeding. And maybe a few medical chems both helpful and malicious.
